### PR TITLE
Split Azure interface into service and customer ones

### PR DESCRIPTION
**Pull request by lzap**
It's cleaner to have service operations within a type that can be reused. It introduces another interface and a second Get function, but it is only for service account operations so it will likely be small.

We should do the same for EC2 and GCP.

### CODE REVIEW

File: internal/clients/http/azure/azure_client.go [hunk 1]
```diff
@@ -10,61 +10,55 @@ import (
...
-
-func newAzureClient(ctx context.Context, auth *clients.Authentication) (clients.Azure, error) {
+func newCustomerClient(ctx context.Context, auth *clients.Authentication) (clients.Azure, error) {
 	opts := azidentity.ClientSecretCredentialOptions{}
 	identityClient, err := azidentity.NewClientSecretCredential(config.Azure.TenantID, config.Azure.ClientID, config.Azure.ClientSecret, &opts)
 	if err != nil {
 		return nil, fmt.Errorf("unable to init Azure credentials: %w", err)
 	}
 
-	return &azureClient{
```
Comment (adiabramovitch) [inline internal/clients/http/azure/azure_client.go:37]: I would keep azureClient for consistency or change in ec2 and GCP to client as well

File: cmd/typesctl/providers/azure_types.go [hunk 2]
```diff
@@ -151,13 +50,5 @@ func generateTypesAzure() error {
 		return fmt.Errorf("unable to generate types: %w", err)
 	}
```
Comment (adiabramovitch) [inline cmd/typesctl/providers/azure_types.go:52]: Why we don't want it as part of the interface as well?

File: internal/clients/interface.go [hunk 3]
```diff
@@ -75,13 +75,20 @@ type EC2 interface {
...
+var GetCustomerAzureClient func(ctx context.Context, auth *Authentication) (Azure, error)
+
+// GetServiceAzureClient returns an Azure client for the service account itself.
+var GetServiceAzureClient func(ctx context.Context) (AzureService, error)
 
 type Azure interface {
 	ClientStatuser
 }
 
+type AzureService interface {
```
[the remark this answers is not in the document]
Comment (lzap) [reply internal/clients/interface.go:88]: Oh yeah it is AzureCustomer vs AzureService. We might need to find a better name, granted, it reads weird.

File: internal/clients/http/azure/azure_client.go [hunk 4]
```diff
@@ -10,61 +10,55 @@ import (
...
 
-type azureClient struct {
+type client struct {
 	subscriptionID string
 	credential     *azidentity.ClientSecretCredential
 }
 
 func init() {
-	clients.GetAzureClient = newAzureClient
+	clients.GetCustomerAzureClient = newCustomerClient
```
[the remark this answers is not in the document]
Comment (lzap) [reply internal/clients/http/azure/azure_client.go:21]: Because of the two different (clients) types. For Azure it is literally a different type, for EC2 it can be the same client (with just different configuration), is it is technically possible to have a single interface for EC2, but for consistency, I would recommend to do the same approach for all the three.

File: internal/clients/http/azure/azure_client.go (hunk 1 above)
Comment (lzap) [reply internal/clients/http/azure/azure_client.go:37]: Because each implementation lives its in own package, you can just use `client` and there will be no conflicts. Whatever you like.

File: cmd/typesctl/providers/azure_types.go (hunk 2 above)
Comment (lzap) [reply cmd/typesctl/providers/azure_types.go:52]: The `typesctl` is a quick and dirty code, it uses three function pointers, feel free to refactor this if you want.

File: internal/clients/http/azure/azure_service_client.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to init Azure credentials: %w", err)
+	}
+
+	return &serviceClient{
+		credential: identityClient,
+	}, nil
+}
+
+func (c *serviceClient) RegisterInstanceTypes(ctx context.Context, instanceTypes *clients.RegisteredInstanceTypes, regionalTypes *clients.RegionalTypeAvailability) error {
```
Comment (avitova) [inline internal/clients/http/azure/azure_service_client.go:35]: This function is 90 lines long. I suggest splitting this, there are some logic parts that can be separated. Especially the huge for loop. Moreover, on the line 118, the for loop is nested on level four.
We could have a helper function to return `instanceType.Cores` value. *c.name comparisons seem like a good example for decomposition as well.

File: internal/clients/http/azure/azure_service_client.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+					}
+					instanceType.MemoryMiB = int64(memoryGB * 1000)
+				}
+				if *c.Name == "HyperVGenerations" {
+					if strings.Contains(*c.Value, "V1") {
+						instanceType.AzureDetail.GenV1 = true
+					}
+					if strings.Contains(*c.Value, "V2") {
+						instanceType.AzureDetail.GenV2 = true
+					}
```
Comment (avitova) [inline internal/clients/http/azure/azure_service_client.go:99]:
```suggestion
					instanceType.AzureDetail.GenV1 = strings.Contains(*c.Value, "V1")
					instanceType.AzureDetail.GenV2 = strings.Contains(*c.Value, "V2") 
```

File: internal/clients/http/azure/azure_service_client.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+			}
+			if vcpusPerCore != 0 {
+				instanceType.Cores = instanceType.VCPUs / vcpusPerCore
+			} else {
+				// Some types have no HT (HPC AMD CPUs) thus vcpus per core is zero:
+				// Standard_B16ms, Standard_B20ms, Standard_HB60rs, Standard_HC44rs,
+				// Standard_M208ms_v2, Standard_M208s_v2, Standard_M416ms_v2,
+				// Standard_M416s_v2 and Standard_PB6s
+				instanceType.Cores = instanceType.VCPUs
+			}
```
Comment (avitova) [inline internal/clients/http/azure/azure_service_client.go:110]:
```suggestion
			instanceType.Cores = instanceType.VCPUs
			if vcpusPerCore != 0 {
				instanceType.Cores /= vcpusPerCore
			}
```

This is also possible, but if we need the comments, a helper function might also come in handy.

File: internal/clients/http/azure/azure_service_client.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+			vcpusPerCore := int32(0)
+			for _, c := range v.Capabilities {
+				if *c.Name == "CpuArchitectureType" {
+					instanceType.Architecture, pageErr = clients.MapArchitectures(ctx, *c.Value)
+					if pageErr != nil {
+						return fmt.Errorf("unable to generate types: %w", pageErr)
+					}
+				}
+				if *c.Name == "vCPUs" {
```
Comment (avitova) [inline internal/clients/http/azure/azure_service_client.go:64]: I expect the *c.Name to have only one of these values, should we maybe use `else if` just for the better readability?

File: internal/clients/interface.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adiabramovitch) [reply internal/clients/interface.go:88]: reminder: we decided to change it to ServiceAzure and keep Azure

File: internal/clients/http/azure/azure_service_client.go (hunk 8 above)
Comment (adiabramovitch) [reply internal/clients/http/azure/azure_service_client.go:64]: Maybe a switch case of c.Name?

File: internal/clients/http/azure/azure_service_client.go (hunk 5 above)
Comment (lzap) [reply internal/clients/http/azure/azure_service_client.go:35]: That is a fair point, I extracted a new function called `typeFromSKU` that moves probably half of the function into a separate one. That cut's it at half, a good starting point I guess.

Unfortunately, I accidentally merged it into the same commit, this should have been done in a separate commit, or separate PR ideally. Sorry for that, it is gonna be hard to follow the change now since the code moved from a different file, but it was essentially just extraction using GoLand "extract method" feature plus I renamed variable `v` to `resourceSKU` which is much more appropriate.

File: internal/clients/http/azure/azure_service_client.go (hunk 6 above)
Comment (lzap) [reply internal/clients/http/azure/azure_service_client.go:99]: Damn, you really hate these, do you? Right, nice point.

File: internal/clients/http/azure/azure_service_client.go (hunk 7 above)
Comment (lzap) [reply internal/clients/http/azure/azure_service_client.go:110]: Sure, I shortened the comment.

File: internal/clients/http/azure/azure_service_client.go (hunk 8 above)
Comment (lzap) [reply internal/clients/http/azure/azure_service_client.go:64]: Switch it is.

**Comment by lzap** [comment]
Done, thanks for comments. Again, the type refresh is in a separate commit so make sure to just look on the code change one.

File: internal/clients/http/azure/azure_client.go [hunk 9]
```diff
@@ -10,61 +10,55 @@ import (
...
 
-type azureClient struct {
+type client struct {
 	subscriptionID string
 	credential     *azidentity.ClientSecretCredential
 }
 
 func init() {
-	clients.GetAzureClient = newAzureClient
+	clients.GetAzureClient = newCustomerClient
```
Comment (adiabramovitch) [inline internal/clients/http/azure/azure_client.go:21]: In GCP(and aws) the customer function is named newGCPClient
(Without a change) can we be consistent?

**Review by adiabramovitch — APPROVED** [review]
I am good here, just a nit. 
@avitova what about you?

File: internal/clients/http/azure/azure_client.go (hunk 9 above)
Comment (ezr-ondrej) [reply internal/clients/http/azure/azure_client.go:21]: to add my 2c I like `newAzureClient` more, as Customer might be missleading, if this service were to be used outside of it's the RH context

Comment (adiabramovitch) [reply internal/clients/http/azure/azure_client.go:21]: @lzap Can we keep newAzureClient?

Comment (lzap) [reply internal/clients/http/azure/azure_client.go:21]: Absolutely, adding an extra commit.

**Comment by lzap** [comment]
New commit is up.

**Comment by adiabramovitch** [comment]
@lzap  Thanks :)
